add-reviews

Nucleus1337 opened the pull request:
feat: добавить функциональность "Отзывы";

Nucleus1337 committed:
add-reviews
feat: добавить функциональность "Отзывы";

Alex-Naumenko1986 commented on src/main/java/ru/yandex/practicum/filmorate/repository/db/ReviewDbStorage.java:
```diff
@@ -0,0 +1,156 @@
+package ru.yandex.practicum.filmorate.repository.db;
```
Тесты не писал для своего решения? Я просто помню, что на созвоне хотели покрывать решения тестами.

Nucleus1337 replied:
Разве? Мне показалось, что наоборот решили ограничиться только постманом..

Alex-Naumenko1986 commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,84 @@
+        filmStorage.getById(userId);
+        log.info("Проверяем существование фильма с ID = {}", userId);
+        userStorage.getById(filmId);
+    }
```
У filmStorage запрашивается id юзера, а у userStorage запрашивается id фильма. Лучше сделать наоборот.

Nucleus1337 replied:
Исправил

Alex-Naumenko1986 commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,84 @@
+    public Review findById(Integer id) {
+        try {
+            log.info("Ищем отзыв с ID = {}", id);
+            return reviewStorage.findById(id);
```
Логи в проекте Евгения на английском, предлагаю для единообразия все делать на английском

Nucleus1337 replied:
Исправил

Alex-Naumenko1986 commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,84 @@
+            return reviewStorage.findById(id);
+        } catch (EmptyResultDataAccessException e) {
+            throw new NotFoundException(String.format("Отзыв с ID = %s не существует", id));
+        }
```
Сообщения исключений в проекте Евгения на английском, поэтому лучше для единообразия тоже сделать на английском

Nucleus1337 replied:
Исправил

Alex-Naumenko1986 commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,84 @@
+    public Review addLike(Integer id, Integer userId, int grade) {
+        log.info("Пользователь ID = {} отгеагировал ({}) на отзыв ID = {}", userId, grade, id);
+        return reviewStorage.addLike(id, userId, grade);
+    }
```
Если попытаться добавить лайк к отзыву с несуществующим id, или отзыв от юзера с несуществующим id, то в ответ приходит ошибка PreparedStatementCallback; SQL. Предлагаю перед добавлением лайка проверить, что такой отзыв и такой пользователь существуют.

Nucleus1337 replied:
Добавил

Alex-Naumenko1986 commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,84 @@
+        log.info("Удаляем отзыв ID = {} от пользователя ID = {}", id, userId);
+        return reviewStorage.removeLike(id, userId);
+    }
+
```
Если попытаться удалить лайк на отзыв с несуществующим id, или от пользователя с несуществующим id, то в ответ приходит непонятная ошибка. Лучше перед удалением лайка проверить, что такой пользователь и такой отзыв существуют.

Nucleus1337 replied:
Добавил

Alex-Naumenko1986 reviewed (commented):
В целом все отлично, оставил несколько замечаний

Nucleus1337 committed:
Merge branch 'develop' into add-reviews

# Conflicts:
#	src/main/resources/schema.sql

Nucleus1337 committed:
add-reviews
feat: добавить проверки, что пользователь или отзыв существует на операции "Добавление лайка", "Удаление лайка", "Удаление отзыва"
fix: поменять язык логов; исправить проверки пользователей и фильмов (перепутаны ID)

AndreiSmazin commented on src/main/java/ru/yandex/practicum/filmorate/repository/db/ReviewDbStorage.java:
```diff
@@ -0,0 +1,156 @@
+                review.getIsPositive(),
+                review.getReviewId());
+
+        return findById(review.getReviewId());
```
Мне кажется это лишнее. Метод можно сделать void. Ты же findById для проверки всё-равно вызываешь в сервисе. А так получается лишнее обращение к базе

Nucleus1337 replied:
Нет, void его делать точно нельзя, если бороться за каждый запрос к БД. Если будет несуществующий ID, то с void не будет понятно, обновилось что-то или нет. Падать будет только на момент, когда надо будет получить обновленный отзыв.
Переделаю несколько.

AndreiSmazin commented on src/main/java/ru/yandex/practicum/filmorate/repository/db/ReviewDbStorage.java:
```diff
@@ -0,0 +1,156 @@
+                userId,
+                grade);
+
+        return findById(id);
```
см. выше

Nucleus1337 replied:
Тут не получится без проверки. Или делать вот такую проверку, как сейчас. Или нужно парсить руками текст ошибки, чтобы выводить в консоль или в json-ответ не "_Нарушение ссылочной целостности: "REVIEWS_LIKES_FK: PUBLIC.REVIEWS_LIKES FOREIGN KEY(REVIEW_ID) REFERENCES PUBLIC.REVIEWS(ID) (2)_", а что-то более человека-читаемое

AndreiSmazin commented on src/main/java/ru/yandex/practicum/filmorate/repository/db/ReviewDbStorage.java:
```diff
@@ -0,0 +1,156 @@
+                id,
+                userId);
+
+        return findById(id);
```
и тут см. выше

Nucleus1337 replied:
В данном варианте, казалось бы, что можно переделать, как в твоем первом комментарии, но так как удаление происходит по 2м параметрам, то сказать, какой из них неверный (или быть может оба) невозможно.
Можно попробовать переделать на ситуативную проверку. То есть, если удалено 0 строк, то только тогда уже делать проверку на существование конкретных ID

AndreiSmazin commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,102 @@
+        }
+    }
+
+    public List<Review> findAllOrByFilmId(String id, String count) {
```
А почему String? Можно же Integer. он тоже будет null если не указать id. И парсить его не надо

Nucleus1337 replied:
Да, точно. Надо будет переделать

AndreiSmazin commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,102 @@
+        reviewStorage.removeById(id);
+    }
+
+    private void checkUserAndLikeExists(Integer reviewId, Integer userId) {
```
Для этого метода логи не нужны, на мой взгляд. Достаточно будет инфы от методов поиска которые ты вызываешь

Nucleus1337 replied:
Лишние логи уберу.

Nucleus1337 committed:
add-reviews
fix: убрать избыточный комментарий в методе checkUserAndLikeExists;
refactor: вызов проверок делать ситуативно для методов removeLike, removeById, update; для метода findAllOrByFilmId поменять тип  параметров String -> Integer

AndreiSmazin commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -50,24 +50,23 @@ public Review findById(Integer id) {
+            throw new NotFoundException(
+                    String.format("REVIEW with ID = %s does not exists", review.getReviewId()));
+        }
+        return findById(review.getReviewId());
```
А почему нельзя просто вернуть review, которое пришло от контроллера?

Nucleus1337 replied:
У отзыва есть расчетное значение "полезности" зависящее от пользователей. Обновлять это значение из объекта не представляется реальным, так как передается только значение, а не зависимости, кто и что поставил. Потому я и тяну весь отзыв целиком из базы

AndreiSmazin commented on src/main/java/ru/yandex/practicum/filmorate/repository/db/ReviewDbStorage.java:
```diff
@@ -48,23 +48,21 @@ public Review add(Review review) {
-    public void removeById(Integer id) {
-        jdbcTemplate.update("delete from reviews where id = ?", id);
+    public int removeById(Integer id) {
+        return jdbcTemplate.update("delete from reviews where id = ?", id);
```
Если я не ошибаюсь такой запрос всегда возвращает 0, и впринципе при удалении строки из базы обратную связь не получить. Так этот метод всегда будет 0 возвращать. Его можно тупо void сделать

Nucleus1337 replied:
В документации к методу сказано:
> Returns:
> the number of rows affected

Могу добавить логирование того на сколько строк повлиял вызов метода update

AndreiSmazin commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -78,25 +77,23 @@ public Review addLike(Integer id, Integer userId, int grade) {
-
         log.info("Remove REVIEW with ID = {} from USER with ID = {}", id, userId);
-        return reviewStorage.removeLike(id, userId);
+        if (reviewStorage.removeLike(id, userId) == 0) {
```
А почему не сделал просто  checkUserAndLikeExists а потом reviewStorage.removeLike, без условия?

Nucleus1337 replied:
Потому что из твоего предыдущего комментария я понял так, что тебе не нравится 3 обращения к базе в этом методе :) Потому мы делаем проверку только в том случае, если у нас ничего не удалилось, а значит какой-то из переданных параметров (userId или reviewId) оказались неверны.

AndreiSmazin commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -78,25 +77,23 @@ public Review addLike(Integer id, Integer userId, int grade) {
-
         log.info("Remove REVIEW with ID = {}", id);
-        reviewStorage.removeById(id);
+        if (reviewStorage.removeById(id) == 0) {
```
Тут аналогично тому что выше

Nucleus1337 replied:
Здесь проверка для того, чтобы понять, а удалилось ли у нас вообще хоть что-то

AndreiSmazin reviewed (approved):
Короче ты прав. Вот тебе апрув. Отличная работа)
